# Analysis-7 Final Analysis

You’ve had the opportunity to analyze two datasets thus far; now it’s time to analyze a dataset of your own. Find a dataset online and use Spark (or Hadoop) to analyze it. You should:

1. Describe the dataset (Earthquack)
2. Outline the types of insights you hope to gain from it
3. Make hypotheses about what you might find
4. Design at least 3 “questions” (along the lines of those above) and answer them. Remember that presentation matters here. ML Models are a good choice for some of the datasets; you can describe what you’ll try to predict or classify and outline your experiences with various models.

Q1) Howmany people talk in subreddit earthquake and the correlation on a real earthquake

Q2) 10 years get data from Earthquake web site. 1990=2310, ,  2006=190, 2007=200, 2008=230  , 2021=Make Prediction.

Q3) 


# Prediction About Upcoming New Reddit User Numbers (Apply ML Model) 

Based on the given data set, we will find the unique user numbers in reddit for each month and then we will make prediction for upcoming month's unique user numbers. (We will use ML models) By doing this reddit development team can prepare themselves for upcoming load.

In [1]:
%%time
spark.catalog.dropGlobalTempView("Comments")
df = sqlContext.read.json("hdfs://orion11:11001/sampled_reddit_v3/*")
df.createGlobalTempView("Comments")

CPU times: user 5 ms, sys: 5.02 ms, total: 10 ms
Wall time: 23.8 s


In [2]:
%%time
botExpr = "[bB][oO][tT]"

filteredComment = (df
                   .filter(~(df.body.like("[deleted]") 
                             | df.body.like('[removed]') 
                             | df.author.rlike(botExpr) 
                             | df.author.like("[deleted]")
                            )
                          )
                  )
df.unpersist()
filteredComment.cache()
print(filteredComment.count())

2742841
CPU times: user 11.7 ms, sys: 2.55 ms, total: 14.2 ms
Wall time: 34.7 s


### Add month and year column into the dataset

In [3]:
%%time
from pyspark.sql.functions import year, month, dayofmonth, from_unixtime
from pyspark.sql.types import DateType

filteredComment = (filteredComment
      .withColumn("year", year(from_unixtime("created_utc").cast(DateType())))
      .withColumn("month",month(from_unixtime("created_utc").cast(DateType()))))
filteredComment.count()

CPU times: user 13.4 ms, sys: 3.09 ms, total: 16.5 ms
Wall time: 1.19 s


2742841

### Q1) Correlation between people comment in subreddit about earthquak and actual earthquake per year: 

Here we have targetted the US regien as shown bellow and use the 

In [2]:
%%time
originDF = sqlContext.read.json("hdfs://orion11:11001/sampled_reddit_v3/*")

CPU times: user 3.73 ms, sys: 77 µs, total: 3.81 ms
Wall time: 3.37 s


In [8]:
# SAMPLED Data
import pyspark.sql.functions as func

#df = sqlContext.read.json("hdfs://orion11:41001/sampled_reddit/*")
#df.createOrReplaceTempView("comments")
# df.saveAsTable("Unique subreddit comments")

# subredditDF = filteredComment.select("subreddit").distinct()

subredditDf = filteredComment.groupBy('year','subreddit').agg(func.count(func.lit(1)))
subredditDf.show()
subredditDf = (subredditDf
               .filter(subredditDf.subreddit.rlike('earthquake'))
               .groupBy('year')
               .agg(func.count(func.lit(1)).alias('NumofSubReddit')))

subredditDf.show()


# subredditDF = SubrredditDF.filterBy('subredd')(subreddit
#                         .groupBy('year')
#                         .agg(func.count('Earthquake').alias('numOfComments')))


# sqlDF=spark.sql("SELECT count(distinct subreddit) AS Unique_Subreddit from comments")
# sqlDF.show()



# df = sqlContext.read.json("hdfs://orion11:41001/reddit/*")
# df.createOrReplaceTempView("Comments")

# Read from 2005 to 2017

# sqlDF=spark.sql("SELECT subredit,count(*) AS CommentCount from Comments and subreddit='Earthquake' and (created_utc>=1104552009000 and created_utc<=1328054399)")

# sqlDF.show()
# sqlDF.show(df.count())


+----+--------------------+--------+
|year|           subreddit|count(1)|
+----+--------------------+--------+
|2014|          skyrimmods|     136|
|2014|                deaf|       8|
|2014|               dubai|      30|
|2014|                lost|      20|
|2014|        raspberry_pi|      59|
|2014|terriblefacebookm...|      16|
|2014|       ClickerHeroes|      89|
|2014|            GameSale|      52|
|2014|RandomActsOfChris...|      14|
|2014|         RateMyMayor|      90|
|2014|              Flyers|      94|
|2014|          EmeraldPS2|      19|
|2014|     douglovesmovies|       5|
|2015|               trees|    2034|
|2015|      TumblrInAction|    1815|
|2014|          glassheads|      30|
|2015|           firewater|       6|
|2015|            formula1|     842|
|2014|              ukguns|       5|
|2015|              Nissan|      10|
+----+--------------------+--------+
only showing top 20 rows

+----+--------------+
|year|NumofSubReddit|
+----+--------------+
+----+--------------

In [10]:
earthquakeComment = filteredComment.filter(filteredComment.body.rlike('earthquake'))
countDF = earthquakeComment.groupBy('year').agg(func.count(func.lit(1))).sort(func.desc('year'))
countDF.show()

+----+--------+
|year|count(1)|
+----+--------+
|2017|      19|
|2016|      95|
|2015|      84|
|2014|      59|
|2013|      39|
|2012|      40|
|2011|      21|
|2010|       8|
|2009|       1|
|2008|       2|
+----+--------+



In [11]:
countDF.coalesce(1).write.format('csv').save('hdfs://orion11:41001/A7')

In [15]:
from pyspark.sql import SQLContext
from pyspark.sql.types import *
sqlContext = SQLContext(sc)
df = sqlContext.read.format('com.databricks.spark.csv').options(header='true').load('hdfs://orion11:41001/Earthquake/2017.csv')
df.count()

3228

![Q1 - Analysis](earthquake1.png)

In [26]:
%%time
import pandas as pd
    
# Load data
earthquake_file_path = '/home4/rteymourzadeh/Project3BigData/P3-p3-hiep_alper_team/A7_Final_Analysis/ML.csv'
earthquake_data = pd.read_csv(earthquake_file_path) 
earthquake_data.columns

# Choose target and features
y = earthquake_data.Earthquake
earthquake_features =['Year']

X = earthquake_data[earthquake_features]

from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
# The split is based on a random number generator. Supplying a numeric value to
# the random_state argument guarantees we get the same split every time we
# run this script.
train_X, val_X, train_y, val_y = train_test_split(X, y,random_state = 0)

CPU times: user 3.87 ms, sys: 970 µs, total: 4.84 ms
Wall time: 8.24 ms


In [27]:
%%time
#ML.csv (Year,earthquake)
import numpy as np
np_arr1 = np.array([[2018],[2019],[2020]])


from sklearn.linear_model import LinearRegression
#from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

#forest_model = RandomForestRegressor(n_estimators=100,random_state=1)
forest_model = LinearRegression()

forest_model.fit(train_X, train_y)
#forest_model.fit(X,y)

val_X=np.append(val_X,np_arr1,axis=0)  
print(val_X)


#val_X=val_X.insert([4],[2017])
melb_preds = forest_model.predict(val_X)
print(melb_preds)
#print('Mean_absolute_error:',mean_absolute_error(val_y, melb_preds))
#print(val_X.dtypes)

[[2012]
 [2017]
 [2010]
 [2018]
 [2019]
 [2020]]
[2723.9537037  4098.16666667 2174.26851852 4373.00925926 4647.85185185
 4922.69444444]
CPU times: user 2.76 ms, sys: 0 ns, total: 2.76 ms
Wall time: 1.74 ms


### Result and Discussion

We tried to calculate the number of earthquakes that may happened in 2018, 2019 and 2020. Those values are: 

Colons can be used to align columns.

#### Trained Data:

**No** | **Year** | **Earthquake**
--- | --- | ---
1 | 2018 | 4373
2 | 2019 | 4647
3 | 2020 | 4922

Since we have a real data from the data set for 2018 and 2019 we can see the accuracy of the model: 

#### Real Data:

**No** | **Year** | **Earthquake**
--- | --- | ---
1 | 2018 | 2135
2 | 2019 | 4152
3 | 2020 | ?



![Q2 - Analysis](earthquake3.png)
